### PR TITLE
Add options for sparse checkout in GitFetcher

### PR DESCRIPTION
**Pull request by psakievich**
Newer versions of git have a beta feature called sparse checkout that allow users to check out a portion of a large repo.

This feature will be ideal for monolithic repo projects that want to model their infrastructure via spack.  This PR implements an addition to the GitFetcher that allows users to add a `git_sparse_paths` attribute to package classes or versions which will then use sparse checkout on those directories/files for the package.

Steps for completion

- [x] Add in `git_sparse_paths` attribute and test it can be called for versions and package attributes
- [x] Update git fetch calls to operate with sparse fetch under these conditions
- [x] Update unit-test infrastructure to handle case with sparse checkout
- [x] Test on a live application
- [x] Update documentation

This PR also refactors the `GitFetcher.clone` method to break it up into somewhat smaller functions.  This is to improve readability and  compatibility with further refactors with #44319 when we will be able to do the vast majority of git clones per commit instead per branch/tag.

### CODE REVIEW

**Comment by psakievich** [comment]
@spackbot fix-style

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by mdmosby** [comment]
We are modeling the internal products in the SIERRA code suite as spack packages and this sparse checkout feature is reducing the fetch times from O(min) to O(s) **for each package** and there are O(50) packages and growing (we are actively converting things). This is a huge win for us in monorepo land!

File: lib/spack/docs/packaging_guide.rst [hunk 1]
```diff
@@ -1361,6 +1366,41 @@ Submodules
   For more information about git submodules see the manpage of git: ``man
   git-submodule``.
 
+Sparse-Checkout
+  You can supply ``git_sparse_paths`` at the package or version level to utilize git's 
+  sparse-checkout feature. This will only clone the paths that are specified in the 
+  ``git_sparse_paths`` attribute for the package along with the files in the top level directory.
+  This feature allows you to only clone what you need from large repository.
```
Comment (mdmosby) [inline lib/spack/docs/packaging_guide.rst]:
```suggestion
  This feature allows you to only clone what you need from a large repository.
```

File: lib/spack/spack/fetch_strategy.py [hunk 2]
```diff
@@ -917,6 +931,80 @@ def clone(self, dest=None, commit=None, branch=None, tag=None, bare=False):
...
+
+        if self.git_version < spack.version.Version("2.25.0.0"):
+            tty.warn(
+                (
+                    f"{self.package.name} is configured for git sparse-checkout "
+                    "but the git version is too old to support sparse cloning. "
+                    "Cloning the full repository instead."
+                )
+            )
+            self.clone_src(commit, branch, tag)
```
Comment (becker33) [inline lib/spack/spack/fetch_strategy.py]:
```suggestion
            self._clone_src(commit, branch, tag)
```
Can you add a test that covers this branch as well?

File: lib/spack/spack/fetch_strategy.py [hunk 3]
```diff
@@ -917,6 +931,80 @@ def clone(self, dest=None, commit=None, branch=None, tag=None, bare=False):
...
+        else:
+            # default to depth=2 to allow for retention of some git properties
+            depth = kwargs.get("depth", 2)
```
Comment (becker33) [inline lib/spack/spack/fetch_strategy.py:958]: Not necessarily for this PR, but do you believe we should be applying this everywhere through `_clone_src` as well?

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/fetch_strategy.py (hunk 3 above)
Comment (psakievich) [reply lib/spack/spack/fetch_strategy.py:958]: 🤷🏻  I don't have strong feelings. I forgot I put this here. I think it was your suggestion from slack... 😜

File: lib/spack/spack/fetch_strategy.py (hunk 2 above)
Comment (psakievich) [reply lib/spack/spack/fetch_strategy.py]: What would you like tested? This is the original code path so it should be covered by all the other tests using `mock_git_repository`.

File: lib/spack/spack/fetch_strategy.py (hunk 3 above)
Comment (becker33) [reply lib/spack/spack/fetch_strategy.py:958]: Yeah, let's plan for future work to apply it everywhere unless someone disagrees.

File: lib/spack/spack/fetch_strategy.py (hunk 2 above)
Comment (becker33) [reply lib/spack/spack/fetch_strategy.py]: A test of using a package that supports sparse checkout with a git version that does not, so we test that the fallback works properly.

**Comment by psakievich** [comment]
@spackbot fix style

**Review by becker33 — APPROVED** [review]

**Pull request merged**